osquery/tables/system/linux/pci_devices.cpp

```diff
@@ -237,6 +238,143 @@ Status PciDB::getSubsystemInfo(const std::string& vendor_id,
   return Status::success();
 }
 
+Status splitVendorModelAttrs(std::string pci_id_attr,
+                             std::string& vendor,
+                             std::string& model) {
+  // pci.ids lower cases everything, so we follow suit.
+  boost::algorithm::to_lower(pci_id_attr);
+  std::vector<std::string> ids;
+  boost::split(ids, pci_id_attr, boost::is_any_of(":"));
+
+  if (ids.size() != 2) {
+    return Status::failure(
+        "Expected 2 identifiers from sysFs PCI device attribute, but "
+        "got " +
+        std::to_string(ids.size()));
+  }
+
+  vendor = std::move(ids[0]);
+  model = std::move(ids[1]);
+  return Status::success();
+}
+
+void extractOEMVendorModelFromPciDB(Row& row,
+                                    const std::string& vendor_id,
+                                    const std::string& model_id,
+                                    const PciDB& pcidb) {
+  row.emplace("vendor_id", "0x" + vendor_id);
+  row.emplace("model_id", "0x" + model_id);
+
+  std::string content;
+  if (pcidb.getVendorName(vendor_id, content).ok()) {
+    row["vendor"] = std::move(content);
+  }
+
+  content.clear();
+  if (pcidb.getModel(vendor_id, model_id, content).ok()) {
+    row["model"] = std::move(content);
+  }
+}
+
+void extractSubsysVendorModelFromPciDB(Row& row,
+                                       const std::string& vendor_id,
+                                       const std::string& model_id,
+                                       const std::string& subsys_vendor_id,
+                                       const std::string& subsys_model_id,
+                                       const PciDB& pcidb) {
+  row.emplace("subsystem_vendor_id", "0x" + subsys_vendor_id);
+  row.emplace("subsystem_model_id", "0x" + subsys_model_id);
+
+  std::string content;
+  if (pcidb.getVendorName(subsys_vendor_id, content).ok()) {
+    row.emplace("subsystem_vendor", std::move(content));
+  }
+
+  content.clear();
+  if (pcidb
+          .getSubsystemInfo(
+              vendor_id, model_id, subsys_vendor_id, subsys_model_id, content)
+          .ok()) {
+    row.emplace("subsystem_model", std::move(content));
+  }
+}
+
+Status extractVendorModelFromPciDBIfPresent(Row& row,
+                                            std::string device_ids_attr,
```

Comment:
I'm curious if this + the subsequent calls to `splitVendorModelAttrs` are creating two copies of the `_addr` strings. I know that `splitVendorModelAttrs` mutates the value passed to it so pass-by-value is correct, but is there an unneeded copy in the local scope here?

Reply:
So the `_ids_attrs` are passed to `extractVendorModelFromPciDBIfPresent` as `rvalue`s from `UdevEventPublisher::getValue` call chaining in lines `348` and `349`, which I thought the compiler optimized so that those contents get moved into `extractVendorModelFromPciDBIfPresent`.  But I guess this is actually quite unclear from the signature, and if the args were passed by `lvalue`, and unnecessary copy would be made.  I can update signature for `extractVendorModelFromPciDBIfPresent` to have the `_ids_attrs` params be `const std::string&`.

Also, I think the uses of `boost::split` in `splitVendorModelAttrs` is a bit more wasteful than it needs to be.  I'll update it to use just `std::find` ...